sdk/storage/azdatalake/sas/service.go

```diff
@@ -0,0 +1,407 @@
+//go:build go1.18
+// +build go1.18
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package sas
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azdatalake/internal/exported"
+)
+
+// DatalakeSignatureValues is used to generate a Shared Access Signature (SAS) for an Azure Storage filesystem or path.
+// For more information on creating service sas, see https://docs.microsoft.com/rest/api/storageservices/constructing-a-service-sas
+// For more information on creating user delegation sas, see https://docs.microsoft.com/rest/api/storageservices/create-user-delegation-sas
+type DatalakeSignatureValues struct {
+	Version              string    `param:"sv"`  // If not specified, this defaults to Version
+	Protocol             Protocol  `param:"spr"` // See the Protocol* constants
+	StartTime            time.Time `param:"st"`  // Not specified if IsZero
+	ExpiryTime           time.Time `param:"se"`  // Not specified if IsZero
+	Permissions          string    `param:"sp"`  // Create by initializing FilesystemPermissions, FilePermissions or DirectoryPermissions and then call String()
+	IPRange              IPRange   `param:"sip"`
+	Identifier           string    `param:"si"`
+	FilesystemName       string
+	FilePath             string // Use "" to create a Filesystem SAS
```

Comment:
this isn't totally accurate right? for Filesystem SAS don't FilePath and DirectoryPath need to be ""

Reply:
that's correct I'll leave a comment on dir as well!